### ISSUE TITLE
RFC: make all unit tests independent (and parallel)

### ISSUE THREAD

**Issue by moul** [gnolang/gno#2376]
In Go, unit tests can share state, and developers need to manually reset the state. While we can follow the same approach, I propose a shift to make unit tests more independent by default. This involves running a new fresh `gno.Machine` for each test, ensuring no shared state and allowing tests to run in parallel.

With independent tests, we can safely run all tests in parallel, speeding up the testing process. Additionally, this ensures that `gno test -run XXX` is consistent, with no test dependencies. Another benefit is that it makes it harder to have implicit or hard-to-read dependencies between tests, improving the overall readability and maintainability of the test suite.

A potential need will be to show more examples of "flows" by writing subtests or test suites (a single `TestXXX` with several steps). However, this doesn't seem like a big deal compared to the benefits of having independent and parallel tests. This issue is an RFC, so please share your thoughts and feedback on this proposal.

**Comment by thehowl** [comment]
Agreed; we'll need to move a lot of the logic of testing back into Go but I don't see that as a big problem. Giving all tests a clean slate is important and good to deviate from Go's behaviour, since in Gno's case they are explicitly meant to change the state.

**Comment by moul** [comment]
By the way, I implemented this in the past, but I think it was reverted: https://github.com/gnolang/gno/pull/2975/files#diff-2e39c5441249018f694183a5710aa7adcad09ec17e870507d20a42184b07e1a7

Essentially, it allowed each `func TestXXX` to run in a clean machine. Perhaps we can extract some parts of my previous PR to enable this again.
